bin/get-video-s3-acls.py

```diff
@@ -0,0 +1,377 @@
+        if video_key not in this_s3_bucket_raw_keys_list:
+            this_all_keys_dict[video_key] = [
+                True,  # NZSL PRESENT
+                False,  # S3 Absent
+                gloss_idgloss,
+                gloss_created_at,
+                gloss_id,
+                video_id,
+                gloss_public,
+                video_public,
+            ]
+
+    return this_all_keys_dict
+
+
+# Cases
+# In S3     In NZSL     Action
+#   Is        Not         Delete S3 Object
+#   Is        Is          Update ACL
+#   Not       Is          Review
+def get_action(key_in_nzsl, key_in_s3):
+    if key_in_s3:
+        if key_in_nzsl:
+            return "Update ACL"
+        else:
+            return "Delete S3 Object"
+    return "Review"
+
+
+# Get S3 object's ACL
+def get_s3_canned_acl(video_key):
+    result = aws_cli(
+        [
+            "s3api",
+            "get-object-acl",
+            "--output",
+            "text",
+            "--query",
+            "Grants[*].Permission",
+            "--bucket",
+            AWS_S3_BUCKET,
+            "--key",
+            video_key,
+        ]
+    )
+    acls_grants = result.stdout.strip().split("\t")
+
+    canned_acl = "unknown"
+    if len(acls_grants) > 1:
+        if acls_grants[0] == "FULL_CONTROL" and acls_grants[1] == "READ":
+            canned_acl = "public-read"
+    elif acls_grants[0] == "FULL_CONTROL":
+        canned_acl = "private"
+
+    return canned_acl
+
+
+# Get S3 object's LastModified date/time
+def get_s3_lastmodified(video_key):
+    result = aws_cli(
+        [
+            "s3api",
+            "head-object",
+            "--output",
+            "text",
+            "--query",
+            "LastModified",
+            "--bucket",
+            AWS_S3_BUCKET,
+            "--key",
+            video_key,
+        ]
+    )
+    return result.stdout.strip()
+
+
+def build_csv_header():
+    return CSV_DELIMITER.join(
+        [
+            "S3 Video key",
+            "Sbank Gloss",
+            "Sbank Gloss created at",
+            "S3 LastModified",
+            "S3 Expected Canned ACL",
+            "S3 Actual Canned ACL",
+            "Sbank Gloss ID",
+            "Sbank Video ID",
+            "Sbank Gloss public",
+            "Sbank Video public",
+            "Action",
+        ]
+    )
+
+
+def build_csv_row(
+    video_key,
+    key_in_nzsl=False,
+    key_in_s3=False,
+    gloss_idgloss=None,
+    gloss_created_at=None,
+    gloss_id=None,
+    video_id=None,
+    gloss_public=False,
+    video_public=False,
+):
+    # See signbank/video/models.py, line 59, function set_public_acl()
+    if key_in_nzsl:
+        canned_acl_expected = "public-read" if video_public else "private"
+    else:
+        canned_acl_expected = ""
+
+    if key_in_s3:
+        lastmodified = get_s3_lastmodified(video_key)
+        canned_acl = get_s3_canned_acl(video_key)
+    else:
+        lastmodified = ""
+        canned_acl = ""
```

Comment:
suggestion: you can get rid of these `else`s by initializing the variables first to their default values

Reply:
Happily. But there are schools that will look over their 👓 at you for doing so! 😆

Reply:
Done.